### ISSUE TITLE
Question regarding trading volume

### ISSUE THREAD

**Issue by arklumichael** [ccxt/ccxt#2562]
- OS:Mac
- Programming Language version: Python 3

Hi,

  I am wondering if the ccxt library has past day trading volume for a specific market pair. What I mean is for example, the total trading volume that has been executed on binance for BTC/USD for the past day or some time period. 

   I see that in the manual, under price ticker, there is bidvolume, askvolume, basevolume , and quotevolume. As i understand, bidvolume and askvolume give me the current volume that i can execute at the best bid/ask price. So it is not what I need. What is basevolume and quotevolume exactly? the description is very short.

   Thanks a lot! Also if you know current any examples of getting historical trading volume, please let me know. 

   Really appreciate your help:)

**Comment by arklumichael** [comment]
thanks!

**Comment by gambitov** [comment]
Dear @kroitor, 

First of all I want to thank you for your work, you are very helpful and invaluable.
Second, I'm trying to get volume usiging:

`order = exchange.fetchTicker('BTC/USDT')
print('Asks',order['askVolume'])
print('Bids',order['bidVolume'])`

It looks like an instant volume, is there a way to get overall volume in a lapse time? for example, total volume in a minute, 5 minutes, 15 minutes, and so on.

**Comment by kroitor** [comment]
@gambitov you want either fetchOHLCV or fetchTrades, both are documented in the Manual. I'd highly recommend to read the entire Manual very carefully from top to bottom at least once – it will save your time, really:

- https://github.com/ccxt/ccxt/wiki/Manual#ohlcv-candlestick-charts
- https://github.com/ccxt/ccxt/wiki/Manual#trades-executions-transactions
- https://github.com/ccxt/ccxt/wiki/Manual#pagination
- ...

**Comment by YuriyTigiev** [comment]
What is the optimal solution to calculate the average daily volume for all pairs  (*/BTC)?  I can use fetchOHLCV but it works slowly if fetch all pairs.

**Comment by kroitor** [comment]
@YuriyTigiev to calculate average daily volume over a period of historical data, the only is to fetchOHLCV.

> but it works slowly if fetch all pairs.

Scale your infrastructure to calculate each pair on a separate computer, this way calculating each pair will take a couple of seconds, and if you parallelize your architecture – you can do that for all pairs within a couple of seconds. There's no easy workaround. You just scale N pairs over M machines/IP addresses. That's how most of the companies do it.

**Comment by YuriyTigiev** [comment]
I use asyncio and for */BTC pairs for a period of one month (june). The process took approximately less than three minutes.  Maybe I do something wrong or maybe this performance is maximum what I can have. 

```
	async def FetchAvgVolume(self, params):
		exchange = self.exchange
		market_type = self.market_type

		excludePairs = params['excludePairs']
		includePairs = params['includePairs']

		timeFrame = params['timeFrame']

		sinceDatetime = params['startDatetime']
		sinceTimestamp = exchange.parse8601(sinceDatetime)
		sinceString = dt.datetime.strptime(sinceDatetime, "%Y-%m-%dT%H:%M:%S").strftime("%Y%m%d%H%M%S")

		tillDatetime = params['tillDatetime']
		tillTimestamp = self.exchange.parse8601(tillDatetime)
		tillString = dt.datetime.strptime(tillDatetime, "%Y-%m-%dT%H:%M:%S").strftime("%Y%m%d%H%M%S")

		activeMarkets = await LoadActiveMarkets( exchange, market_type, includePairs, excludePairs)
		
		pairs = list(activeMarkets.keys())
		quotes =  set([pair.split('/')[1] for pair in pairs])

		avgVolumeList = []		
		for quote in quotes:
			pairsFiltered = list(filter(lambda x: x.split('/')[1] == quote, pairs))

			print(f"*** Quote: {quote} ***")

			tasks = ( FetchOHLC(exchange, pair, sinceTimestamp, tillTimestamp, timeFrame, validateTimePeriod=False) for pair in pairs)
			pending = list( tasks )
			resultsList = []
			while pending:
				done, pending = await aio.wait(
								pending, 
								#timeout=30, 
								return_when=aio.FIRST_COMPLETED
					)
				for future in done:
					result = future.result()
					if(len(result) > 0):
						avgVolume = reduce(lambda x, y:  x+y[5], result[1], 0) / ( len(result[1]) if(len(result[1]) > 0) else 0 )
						avgVolumeList.append([result[0], sinceString, tillString, timeFrame, avgVolume])
					print(f"Fetch OHLCV: {result[0]} {dt.datetime.now()}, {len(pairs)-len(pending)}/{len(pairs)}")


			return pd.DataFrame(avgVolumeList, columns=['pair', 'sinceDate', 'tillDate', 'timeFrame', 'avgVolume'])
```

---

```
async def FetchOHLC(exchange, pair, sinceTimestamp, tillTimestamp, timeFrame, validateTimePeriod = True, attempts = 3, sleep = 60):
	networkError = 0
	fetchError = 0

	while networkError < attempts:
		try:

			since = sinceTimestamp

			timeFrameSeconds = ConvertToSeconds(timeFrame)

			timePeriod = (tillTimestamp - sinceTimestamp)//(1000*timeFrameSeconds)

			ohlc = (await exchange.fetchOHLCV(symbol = pair, timeframe = timeFrame, since = since, limit = timePeriod + 1))[-timePeriod:]

			if(validateTimePeriod == True and len(ohlc) != timePeriod):
				fetchError += 1
				logging.warning("Not all data fetched")
				logging.warning("{}: sleep {} seconds".format(sys._getframe().f_code.co_name, sleep))
				time.sleep(sleep) 
					
				if(fetchError >= attempts):
					raise Exception("Not all data fetched. Except {} got {}".format(timePeriod, len(ohlc)))
				else:
					continue

			return (pair, ohlc)

		except ccxt.NetworkError as ex:
			logging.error("{}, {}".format(sys._getframe().f_code.co_name, ex))
			networkError = networkError + 1

			if(networkError >= attempts):
				raise Exception(ex)

			logging.warning("{}: sleep {} seconds".format(sys._getframe().f_code.co_name, sleep))
			time.sleep(sleep) 

		except Exception as ex: 
			logging.error("{}, {}".format(sys._getframe().f_code.co_name, ex))
			raise Exception(ex)
```

**Comment by kroitor** [comment]
@YuriyTigiev if you're using the built-in rate limiter (or any custom rate limiting), that will take some time – this is just the nature of rate limiting. You can't get everything for all pairs without getting banned. The exchanges won't like that many requests in a split second, therefore those requests have to be stretched over time to stay within the rate limits defined by the exchanges themselves. The only way to workaround that is to fetch half of the pairs from one address, and another half from a different address. That will take 1.5minutes instead of 3 minutes. In other words, you have to parallelize them if you want to speed up the process. However, with parallelization comes a different new task of synchronizing the machines running in parallel.
